Policy parameter update with 2018 values

Peter-Metz opened the pull request:
This PR uses the process described in [Issue #2183](https://github.com/PSLmodels/Tax-Calculator/issues/2183) to update policy parameter values in `policy_current_law.json` to reflect IRS final forms for tax year 2018. After updating final 2018 parameter values, I ran `ppp.py` to calculate parameter values through 2026. Where the IRS forms changed in layout (e.g. Form 1040), I updated the `"irs_ref"` value to point users to the correct line in the 2018 forms.

The IRS forms and instructions used to make these updates can be found below:

[Form 1040](https://www.irs.gov/pub/irs-pdf/f1040.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/i1040gi.pdf), [Form 1040 (Schedule 1)](https://www.irs.gov/pub/irs-pdf/f1040s1.pdf), [Form 1040 (Schedule 2)](https://www.irs.gov/pub/irs-pdf/f1040s2.pdf), [Form 1040 (Schedule 4)](https://www.irs.gov/pub/irs-pdf/f1040s4.pdf), [Form 1040 (Schedule A)](https://www.irs.gov/pub/irs-pdf/f1040sa.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/i1040sca.pdf)

[Form W-2](https://www.irs.gov/pub/irs-pdf/fw2.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/iw2w3.pdf)

[Form 6251](https://www.irs.gov/pub/irs-pdf/f6251.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/i6251.pdf)

[Form 8863](https://www.irs.gov/pub/irs-pdf/f8863.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/i8863.pdf)

[Form 8812](https://www.irs.gov/pub/irs-pdf/f8812.pdf) and [instructions](https://www.irs.gov/pub/irs-pdf/i1040s8.pdf)

cc @martinholmer @MattHJensen

martinholmer commented:
@Peter-Metz,  Thanks for all the work that is included in your pull request #2212.  But as you can see from this

<img width="981" alt="screen shot 2019-01-28 at 5 14 05 pm" src="https://user-images.githubusercontent.com/12170745/51869936-2e0c6280-2320-11e9-95fe-675b4516897e.png">

that there are number of procedural problems.  The biggest procedural problem is that your development work branched off from the master branch **seventeen** days ago just after PR#2185 was merged into the master branch.  (You can see this in the Network time-line graph on GitHub.)  Meanwhile there have been several important revisions to the master branch (including updating the projections in PR#2196 and updating the OASDI maximum taxable earnings parameter in PR#2203).  This means, among other things, that your pull request undoes all the work I did with the `ppp.py` script when @andersonfrailey's new projections were incorporated into the master branch in PR#2196.  The fact that your PR is based on data and code that is out-date can be seen in your personal "origin" repository (where you forked PSLmodels/Tax-Calculator):

<img width="954" alt="screen shot 2019-01-28 at 5 23 33 pm" src="https://user-images.githubusercontent.com/12170745/51870673-4aa99a00-2322-11e9-9c3a-3c618e98f779.png">

Your development branch (which is being submitted  as PR#2212 for incorporation into the PSLmodels/Tax-Calculator master branch) is 7 commits ahead (those are your changes on your `2018-param` development branch) and 71 commits behind the PSLmodels/Tax-Calculator master branch.  

The solution is to merge the most recent version of the PSLmodels/Tax-Calculator master branch into your `2018-param` development branch.  If you're not certain how to do this, it would be good for you to read the [documentation](https://github.com/PSLmodels/Tax-Calculator#tax-calculator) including the [contributor guide](https://github.com/PSLmodels/Tax-Calculator/blob/master/CONTRIBUTING.md#tax-calculator-contributor-guide) and the documentation to which it points.

Peter-Metz commented:
@martinholmer, thank you for the explanation. The latest commit should fix the merge conflicts. The conflicts were results of the numbers generated by `ppp.py`, and did not affect any of the 2018 parameters that I entered. I'm not quite sure why the `ppp.py` script generated different numbers, but if you think the script needs to be re-run, I am happy to do so in a separate PR.

martinholmer commented:
@Peter-Metz said:
> I'm not quite sure why the `ppp.py` script generated different numbers

Because you were using the old projections as I explained earlier in [this comment](https://github.com/PSLmodels/Tax-Calculator/pull/2212#issuecomment-458331728).

Peter-Metz commented:
@martinholmer 

Ah, understood. Thanks.

martinholmer commented:
@Peter-Metz,  I'm curious about why you changed some 2017 policy parameter values as part of your 2018 parameter update work in PR #2212.  For example, you changed (slightly) the values for the `_EITC_ps` and `_EITC_ps_MarriedJ` parameters for 2017.  Was that intentional?  If so, can you provide a link to the IRS documents that show the 2017 values you switched to?

Also, the GitHub diff for the `_II_brk6` parameter is confusing.  What's going on there?

martinholmer commented:
@Peter-Metz, Now that PR #2211 has been merged into the PSLmodels / Tax-Calculator master branch, you need to do two things on your computer in the following order:

1.  On your local master branch, you need to run in the top T-C directory `./gitsync` (or just `gitsync` if working on Windows), and

2. On you local 2018-param branch, you need to run `git merge master`.

Peter-Metz commented:
@martinholmer:

>you need to do two things on your computer in the following order:

Thanks for these instructions.


>you changed (slightly) the values for the _EITC_ps and _EITC_ps_MarriedJ parameters for 2017

This was intentional, but on second thought, I'm not sure I was correct in doing this. For example, I changed 2017, `_EITC_ps` for 0 children from $8,340 to $8,350. The phaseout begins for incomes "at least" $8,350 but "less than" $8,400. I now see that if AGI is being calculated in increments of $10, the original value was correct. If you agree with this assessment, I will push a commit that reverts the 2017 values for `_EITC_ps` and `_EITC_ps_MarriedJ` back to their original values, and fix the 2018 values with that same logic.


>Also, the GitHub diff for the _II_brk6 parameter is confusing. What's going on there?

First, I changed 2018 "widow" from $500,000 to $600,000 because I believe there was a typo. In terms of the values from 2019-2026, that looks like a difference stemming from `ppp.py`, although I'm not sure why that didn't cause a merge conflict. I am happy to revert the 2019-2026 values back or re-run `ppp.py` on my computer now that my development branch is up to date. Either way, if you agree that the 2018 value for "widow" was a typo, `ppp.py` will have to be re-run.

martinholmer commented:
@Peter-Metz said in PR #2212:
>> you changed (slightly) the values for the _EITC_ps and _EITC_ps_MarriedJ parameters for 2017
>
> This was intentional, but on second thought, I'm not sure I was correct in doing this. For example, I changed 2017, _EITC_ps for 0 children from $8,340 to $8,350. The phaseout begins for incomes "at least" $8,350 but "less than" $8,400. I now see that if AGI is being calculated in increments of $10, the original value was correct. If you agree with this assessment, I will push a commit that reverts the 2017 values for _EITC_ps and _EITC_ps_MarriedJ back to their original values, and fix the 2018 values with that same logic.

I agree, so please do these things with the EITC parameters.

>> Also, the GitHub diff for the _II_brk6 parameter is confusing. What's going on there?
>
> First, I changed 2018 "widow" from $500,000 to $600,000 because I believe there was a typo. In terms of the values from 2019-2026, that looks like a difference stemming from ppp.py, although I'm not sure why that didn't cause a merge conflict. I am happy to revert the 2019-2026 values back or re-run ppp.py on my computer now that my development branch is up to date. Either way, if you agree that the 2018 value for "widow" was a typo, ppp.py will have to be re-run.

Which of the IRS forms for which you provide links shows the widow amount at $600,000?

Assuming you have caught a mistake (good work!), the pull request values for _II_brk6 for 2018+ seem just fine.  No need to do anything except verify on the forms that $600,000 is the correct number for widow.

martinholmer commented:
@Peter-Metz,  After fixing the issues discussed in [this comment](https://github.com/PSLmodels/Tax-Calculator/pull/2212#issuecomment-459024285), you need to do one more thing before addressing the test failures.  Change in `policy.py` the value of `LAST_KNOWN_YEAR` from 2017 to 2018.  Then rerun the tests on your computer.  If you're puzzled about how to handle any of the test failures, please ask by posting a comment in this PR conversation.  This PR is shaping up very nicely.  Thanks for the major contribution.

martinholmer commented:
@martinholmer said incorrectly in #2212:
> Which of the IRS forms for which you provide links shows the widow amount at $600,000?
>
Assuming you have caught a mistake (good work!), the pull request values for `_II_brk6` for 2018+ seem just fine. No need to do anything except verify on the forms that $600,000 is the correct number for widow.

Actually, the values currently in #2212 for `_II_brk6` for 2019+ are not fine.  They are apparently based on the use of the old projection factors earlier in the #2212 work.   After correcting the 2018 value for widow there should be differences (relative to the master branch for the widow values.  But that not the case now as shown in the screen shot below (where the master branch is on the left and the development branch is on the right):

<img width="1753" alt="screen shot 2019-01-31 at 10 40 25 am" src="https://user-images.githubusercontent.com/12170745/52065841-5923cb80-2545-11e9-8c61-c11d4a953fe5.png">

What we want in this comparison is the 500000-to-600000 change for widow in 2018, and changes in 2019+ **only** for the widow value.

@Peter-Metz

Peter-Metz commented:
@martinholmer: 

>Which of the IRS forms for which you provide links shows the widow amount at $600,000?

This can be found in the [1040 Instructions](https://www.irs.gov/pub/irs-pdf/i1040gi.pdf), Schedule Y-1, page 113.


>Change in policy.py the value of LAST_KNOWN_YEAR from 2017 to 2018. 

Will do.


>Actually, the values currently in #2212 for _II_brk6 for 2019+ are not fine. 

Thanks, I came to the same conclusion while running the tests and have since fixed this locally by re-running `ppp.py`.


>If you're puzzled about how to handle any of the test failures, please ask by posting a comment in this PR conversation.

I have managed to make the necessary changes to pass most of the tests, but would appreciate some guidance on one failure in particular. The PR fails the function `test_json_file_contents` in the `test_parameters.py` file because some of the `len(rowlabel)` do not match `known_years`. Effectively this test checks that there are as many parameter values as there are known years, in this case through 2018. I am a bit confused by the logic written into the test for certain parameters. See the relevant block of code below:

       # form_parameters are those whose value is available only on IRS form
        form_parameters = ['_AMT_em_pe',
                           '_ETC_pe_Single',
                           '_ETC_pe_Married']
        if param['cpi_inflated']:
            error = False
            known_years = num_known_years
            if pname in long_params:
                known_years = long_known_years
            if pname in form_parameters:
                if len(rowlabel) != (known_years - 1):
                    error = True
            else:
                if pname == '_SS_Earnings_c':
                    if len(rowlabel) != known_years + 2:
                        error = True
                else:
                    if len(rowlabel) != known_years:
                        error = True

If possible, could you please explain why those four parameters (`_AMT_em_pe`,  `_ETC_pe_Single`, `_ETC_pe_Married`, and `_SS_Earnings_c`) have different rules than the other parameters, and what I can do to fix this issue?

martinholmer commented:
@Peter-Metz,  Sounds like you're make great progress and even learning to ignore my advice when its bad advice.

The test that is puzzling you has to do with the fact that in the past a few parameter values were not (as far as we could tell) available in advance documents and had to be updated only after the IRS forms were published.  Hence a few had one less year label.  But in the update you're doing we have all the 2018 values and so there is no need to treat a few parameters differently.  If you set the `form_parameters` list to the empty list, can you pass that test?

Peter-Metz commented:
@martinholmer:

>But in the update you're doing we have all the 2018 values and so there is no need to treat a few parameters differently. If you set the form_parameters list to the empty list, can you pass that test?

Thanks for the explanation. Indeed, when I set the `form_parameters` list to an empty list, the test passes for those three parameters. However, I'm still not sure why the test is looking for `_SS_Earnings_c` values through 2020. As it stands, we have values through 2019. Is it safe to change the test to reflect this?

Another issue came up as I was going through test failures: `_ID_statelocaltax_c` and `_ID_RealEstate_c` , the maximum deduction allowed for state and local income and real estate tax respectively, have not yet been updated to reflect TCJA. As it stands now, values for these two parameters only go through 2017, but as I understand TCJA, the new law limits the deduction of state and local income, sales, and property tax to $10,000 in total (see [1040 Instructions](https://www.irs.gov/pub/irs-pdf/i1040gi.pdf), page 6). Since, in the new law, state/local income and property taxes are combined, but in `policy_current_law.json` the parameters are separate and in `calcfunctions.py` the capped amounts are calculated separately, I suspect that we will need to tweak one or both of those files.

Let me know if you think this observation warrants its own issue and/or PR.

martinholmer commented:
@Peter-Metz said:
> Indeed, when I set the `form_parameters` list to an empty list, the test passes for those three parameters. However, I'm still not sure why the test is looking for `_SS_Earnings_c` values through 2020. As it stands, we have values through 2019. 

Try removing the 2019 year and its value from the `_SS_Earnings_c` record in the `policy_current_law.json` file.  And then look at the failing test to see if you can remove any code that treats the `_SS_Earnings_c` parameter differently than the other policy parameters.

martinholmer commented:
@Peter-Metz said in #2212:
> Another issue came up as I was going through test failures: `_ID_statelocaltax_c` and `_ID_RealEstate_c` , the maximum deduction allowed for state and local income and real estate tax, respectively, have not yet been updated to reflect TCJA.  As it stands now, values for these two parameters only go through 2017, but as I understand TCJA, the new law limits the deduction of state and local income, sales, and property tax to $10,000 in total (see [1040 Instructions](https://www.irs.gov/pub/irs-pdf/i1040gi.pdf), page 6). Since, in the new law, state/local income and property taxes are combined, but in `policy_current_law.json` the parameters are separate and in `calcfunctions.py` the capped amounts are calculated separately, I suspect that we will need to tweak one or both of those files.

Good catch!  I agree that this issue should dealt with separately from this pull request.
Can you open a new issue describing this problem?  Thanks.

But we have to make some kind of change in order to get the tests to pass for pull request #2212 , right?
Seems like we're in a bind here.

martinholmer commented:
@Peter-Metz,  After removing the 2019 year and value for the `_SS_Earnings_c` parameter and hopefully passing that test, can you commit all the local changes on your 2018_params development branch and do a `git push origin 2018_params` command so that we can see the status of PR #2212 on GitHub?

martinholmer commented:
@Peter-Metz,  Actually things are not a bad as they seem with respect to the limitation of itemized deductions of state/local taxes.  The two parameter you mentioned seem not to be used in current (post-TCJA) law.  The parameter doing the limiting seems to be this one:
```
   "_ID_AllTaxes_c": {
        "long_name": "Ceiling on the amount of state and local income, sales and real estate tax deductions allowed (dollars)",
        "description": "The amount of state and local income, sales and real estate tax deductions is limited to this dollar amount.",
        "section_1": "Itemized Deductions",
        "section_2": "State And Local Income And Sales Taxes",
        "irs_ref": "",
        "notes": "",
        "row_var": "FLPDYR",
        "row_label": ["2013",
                      "2014",
                      "2015",
                      "2016",
                      "2017",
                      "2018",
                      "2019",
                      "2020",
                      "2021",
                      "2022",
                      "2023",
                      "2024",
                      "2025",
                      "2026"],
        "start_year": 2013,
        "cpi_inflatable": true,
        "cpi_inflated": true,
        "col_var": "MARS",
        "col_label": ["single", "joint", "separate", "headhousehold", "widow"],
        "boolean_value": false,
        "integer_value": false,
        "value": [[9e99, 9e99, 9e99, 9e99, 9e99],
                  [9e99, 9e99, 9e99, 9e99, 9e99],
                  [9e99, 9e99, 9e99, 9e99, 9e99],
                  [9e99, 9e99, 9e99, 9e99, 9e99],
                  [9e99, 9e99, 9e99, 9e99, 9e99],
                  [10000.0, 10000.0, 5000.0, 10000.0, 10000.0],
                  [10229.0, 10229.0, 5114.5, 10229.0, 10229.0],
                  [10432.56, 10432.56, 5216.28, 10432.56, 10432.56],
                  [10666.25, 10666.25, 5333.12, 10666.25, 10666.25],
                  [10908.37, 10908.37, 5454.19, 10908.37, 10908.37],
                  [11151.63, 11151.63, 5575.81, 11151.63, 11151.63],
                  [11394.73, 11394.73, 5697.37, 11394.73, 11394.73],
                  [11639.72, 11639.72, 5819.86, 11639.72, 11639.72],
                  [9e99, 9e99, 9e99, 9e99, 9e99]],
        "range": {"min": 0, "max": 9e99},
        "out_of_range_minmsg": "",
        "out_of_range_maxmsg": "",
        "out_of_range_action": "stop",
        "compatible_data": {"puf": true, "cps": true}
    },
```
Can you check the code in the `calcfunctions.py` file to make sure that this parameter is being used to do the itemized deduction limitation?  Thanks.

martinholmer commented:
@Peter-Metz,  What about [this comment](https://github.com/PSLmodels/Tax-Calculator/pull/2212#issuecomment-459777015)?  Did you not read it?

Peter-Metz commented:
@martinholmer, I agree with you assessment in [this comment](https://github.com/PSLmodels/Tax-Calculator/pull/2212#issuecomment-459777015) -- that `calcfunctions.py` uses `_ID_AllTaxes_c` to calculate maximum itemized deductions. See the block of relevant code from `calcfunctions.py` below:

    c18400 = min(c18400, ID_StateLocalTax_crt * max(c00100, 0.0001))
    c18500 = min(c18500, ID_RealEstate_crt * max(c00100, 0.0001))
    c18300 = min(c18400 + c18500, ID_AllTaxes_c[MARS - 1])

Still, we need to populate 2018 values for `_ID_StateLocalTax_c` and `_ID_RealEstate_c`. Do you think we should continue populating with `9e99`'s instead of what I have done? My intuition is that the results will be the same, but that may be misguided.

martinholmer commented:
@Peter-Metz, After doing [this](https://github.com/PSLmodels/Tax-Calculator/pull/2212#issuecomment-459857373) don't forget to `./gitsync` on your master branch and then merge the updated master branch into your development branch.  Your development branch is several commits behind the GitHub master branch as can be seen at https://github.com/PSLmodels/Tax-Calculator/network

Peter-Metz commented:
@martinholmer, thanks for the reminder. I have updated `_ID_StateLocalTax_c` and `_ID_RealEstate_c` per the conversation above. As expected, the results did not change.
